python/tk_framework_desktopserver/shotgun/api_v2.py

```diff
@@ -0,0 +1,780 @@
+        :type pc: :class:`~sgtk.bootstrap.ToolkitManager`
+        :param dict project: The project entity.
+
+        :returns: A list of PipelineConfiguration entity dictionaries.
+        :rtype: list
+        """
+        if "pipeline_configurations" not in self.WSS_KEY_CACHE[self._wss_key]:
+            self.WSS_KEY_CACHE[self._wss_key]["pipeline_configurations"] = dict()
+
+        pc_data = self.WSS_KEY_CACHE[self._wss_key]["pipeline_configurations"]
+
+        if project["id"] not in pc_data:
+            pc_data[project["id"]] = manager.get_pipeline_configurations(
+                project=project,
+            )
+        else:
+            self.logger.debug(
+                "Cached PipelineConfiguration entities found for %s" % self._wss_key
+            )
+
+        return pc_data[project["id"]]
+
+    def _get_site_state_data(self):
+        """
+        Gets state-related data for the site. Exactly what data this is depends
+        on the "browser_integration" core hook's "get_site_state_data" method,
+        which returns a list of dicts passed to the Shotgun Python API's find
+        method as kwargs. The data returned by this method is cached based on
+        the WSS connection key provided to the API's constructor at instantiation
+        time. This means that this data is queried from Shotgun only once per
+        unique WSS connection.
+
+        :returns: A list of Shotgun entity dictionaries.
+        :rtype: list
+        """
+        if "site_state_data" not in self.WSS_KEY_CACHE[self._wss_key]:
+            self.WSS_KEY_CACHE[self._wss_key]["site_state_data"] = []
+
+            requested_data_specs = self._engine.sgtk.execute_core_hook_method(
+                "browser_integration",
+                "get_site_state_data",
+            )
+
+            for spec in requested_data_specs:
+                entities = self._engine.shotgun.find(**spec)
+                self.WSS_KEY_CACHE[self._wss_key]["site_state_data"].extend(entities)
+        else:
+            self.logger.debug("Cached site state data found for %s" % self._wss_key)
+
+        return self.WSS_KEY_CACHE[self._wss_key]["site_state_data"]
+
+    def _get_software_entities(self):
+        """
+        Gets all Software entities from the Shotgun client site. Included are
+        all existing fields. This data is cached per WSS connection key that
+        was provided to the API's constructor at instantiation time. This means
+        that this data is queried from SHotgun only once per unique WSS
+        connection.
+
+        :returns: A list of Software entity dictionaries.
+        :rtype: list
+        """
+        if "software_entities" not in self.WSS_KEY_CACHE[self._wss_key]:
+            self.WSS_KEY_CACHE[self._wss_key]["software_entities"] = self._engine.shotgun.find(
+                "Software",
+                [],
+                fields=self._engine.shotgun.schema_field_read("Software").keys(),
```

Comment:
performance: we know the fields, so can we hard code them? if not, please explain in a comment!